OCPBUGS-9070: Fix hotlooping on Cronjob resources

Davoska opened the pull request:
Currently the CVO hotloops on `CronJob` resources.

The issue was caused by the missing defaulting of these resources.

To fix this issue I have added the handling for `CronJob` resources so they are not reconciled as `generic` resources. This has then allowed me to introduce the defaulting.

To fix this issue I have also updated the applying for `Job` resources because its logic can be then reused as part of the fix for the hotlooping on `Cronjob` resources.  While updating the applying for `Job` I have added missing defaulting,  and I have added reconciling of fields that were added later in the development, and the CVO was not reconciling them.

This pull request references https://issues.redhat.com/browse/OCPBUGS-9070

Davoska commented:
/hold
I want to verify in the CI logs that the fix is working as expected.

Davoska commented:
Regarding `bugzilla/invalid-bug`:

Grr, I don't see the `4.14.0` version in the list. I will look into it later.

wking commented on lib/resourcebuilder/resourcebuilder.go:
```diff
@@ -169,6 +169,18 @@ func (b *builder) Do(ctx context.Context) error {
+			updatingMode); err != nil {
+			return err
+		} else if !deleteReq {
+			if _, _, err := resourceapply.ApplyCronJobv1(ctx, b.batchClientv1, typedObject, reconcilingMode); err != nil {
```
One wrinkle with adding official CronJob support is deciding if we want a `return b.checkCronJobHealth(ctx, actual)` like we see below for Jobs.  If we do want that, we can have a second discussion about whether we want it in this pull request, or if we'd rather file a ticket to track it as future work.

petr-muller replied:
Feels like we *should* have it but not necessarily in this PR, just making a note/card should be fine, IMO

Davoska replied:
Good point. I will create a card regarding this issue so we can address it in the future. This pull request already improves the CVO's CronJob support and fixes the addressed issue.

Personally, at first glance, I am not sure how would the `checkCronJobHealth` method look like. And thus it would require additional effort. But as Petr said it does feel like we should have it.

Davoska replied:
I have created [OTA-957: CVO should check the health of CronJob resources](https://issues.redhat.com/browse/OTA-957).

wking commented on lib/resourcemerge/batch.go:
```diff
@@ -2,21 +2,97 @@ package resourcemerge
 func EnsureJob(modified *bool, existing *batchv1.Job, required batchv1.Job) {
-
 	if required.Spec.Selector != nil {
 		panic("spec.selector is not supported in Job manifests")
```
Why doesn't this move down into `ensureJobSpec`?

Davoska replied:
Nice catch, I didn't notice that. I will move it to the `ensureJobSpec`.

petr-muller commented on lib/resourceapply/batch.go:
```diff
@@ -29,16 +29,56 @@ func ApplyJobv1(ctx context.Context, client batchclientv1.JobsGetter, required *
+	if !*modified {
+		return existing, false, nil
+	}
+	if reconciling {
+		if diff := cmp.Diff(&original, existing); diff != "" {
+			klog.V(2).Infof("Updating CronJob %s/%s due to diff: %v", required.Namespace, required.Name, diff)
+		} else {
+			klog.V(2).Infof("Updating CronJob %s/%s with empty diff: possible hotloop after wrong comparison", required.Namespace, required.Name)
+		}
+	}
```
Reading the copy of this code for fifth time makes me think whether we can somehow DRY these `Apply*` methods out :/ Not an actual suggestion for you to fix, just thinking loud.

Davoska replied:
Yes, that would be great :( I will create a card so we can keep track of these potential places to be refactored. This is a good idea for a shift-week activity or for anyone interested in helping to develop the CVO.

petr-muller commented on lib/resourcedelete/batch.go:
```diff
@@ -37,3 +37,32 @@ func DeleteJobv1(ctx context.Context, client batchclientv1.JobsGetter, required
+	if delAnnoFound, err := ValidDeleteAnnotation(required.Annotations); !delAnnoFound || err != nil {
+		return delAnnoFound, err
+	}
+	existing, err := client.CronJobs(required.Namespace).Get(ctx, required.Name, metav1.GetOptions{})
```
Hmmm is it right to entirely ignore errors here? We immediately overwrite it on L56 and we end up using `existing` on L62. Either we should not be ignoring the error, or we should do so explicitly with a comment why (`existing, _ := ...`)

Davoska replied:
It took me few a seconds but the initial `err` does get passed to the function `GetDeleteProgress()` on the L56 and thus is not immediately overwritten:

```go
	if deleteRequested, err := GetDeleteProgress(resource, err); err == nil {
```

And there it is used quite a few times:

https://github.com/openshift/cluster-version-operator/blob/master/lib/resourcedelete/helper.go#L100-L132

Davoska replied:
Sorry, I forgot about your new comment when pushing the new changes. However, the new code structure simplifies the logic in my opinion. 

https://github.com/Davoska/cluster-version-operator/blob/bug-2039228-cvo-hotloops-on-cronjob/lib/resourcedelete/batch.go#L43-L66

```go
	if delAnnoFound, err := ValidDeleteAnnotation(required.Annotations); !delAnnoFound || err != nil {
		return delAnnoFound, err
	}
	existing, err := client.CronJobs(required.Namespace).Get(ctx, required.Name, metav1.GetOptions{})
	resource := Resource{
		Kind:      "cronjob",
		Namespace: required.Namespace,
		Name:      required.Name,
	}
	deleteRequested, err := GetDeleteProgress(resource, err)
	if err != nil {
		return true, fmt.Errorf("Error running delete for %s, err=%v", resource, err)
	}
	if !deleteRequested && updateMode {
		if err := client.CronJobs(required.Namespace).Delete(ctx, required.Name, metav1.DeleteOptions{}); err != nil {
			return true, fmt.Errorf("Delete request for %s failed, err=%v", resource, err)
		}
		SetDeleteRequested(existing, resource)
	}
```

Now it's clearer than before, and something like `getErr` is no longer needed IMO. However, if you feel like it should be there even now, I am happy to add it.

petr-muller replied:
It's fine, no more modifications needed, thanks!

petr-muller commented on lib/resourcedelete/batch.go:
```diff
@@ -37,3 +37,32 @@ func DeleteJobv1(ctx context.Context, client batchclientv1.JobsGetter, required
+		Namespace: required.Namespace,
+		Name:      required.Name,
+	}
+	if deleteRequested, err := GetDeleteProgress(resource, err); err == nil {
```
It'd read a bit better if the condition was inverse:

```go
deleteRequested, err := GetDeleteProgress(resource, err)
if err != nil {
   return true, fmt.Errorf("Error running delete for %s, err=%v", resource, err)
}
if ... {
  if ... {
    return ...
  }
  SetDeleteRequested(...
}
return true, nil
```

Davoska replied:
Me too! We would return early rather than nest deeper.

I'll change the condition in this current PR, and in the future, we can refactor the other conditions in the `resourcedelete` package (or using the DRY principle we will leave reuse this condition).

Davoska commented:
/hold
Oops, wrong push.

Davoska commented:
/unhold

I have addressed the comments in https://github.com/openshift/cluster-version-operator/compare/659889f559b396b5ecb7c37d48aac98557ff0563..f665f3c39136b7e55ee92cb741f9b2d0b4e18a6e.

However, I had to rebase the PR in the latest push.

Davoska commented:
/hold

Non-critical merge

petr-muller reviewed (approved):
:shipit: :tada:

/retest

evakhoni commented:
@Davoska @petr-muller is this one ready for pre-merge verification? any major changes expected?

Davoska commented:
Hello @evakhoni :wave: 

The PR is ready. No new changes are expected.

Davoska commented:
Oh no, it seems I have incorrectly broken up the commits the last week. Now, the refactoring commit https://github.com/openshift/cluster-version-operator/pull/910/commits/2f4e5508c2b103e5381b79d72b1cf128f5a6d54b contains changes that were inside a now non-existing commit that explained the added defaulting for the `Job` resources. Working on it :wrench:

Davoska commented:
Sorry everyone, for the noise! I have fixed the commits, and in the most recent force-push comparison, we can see that no new changes were added to the code.

evakhoni commented:
/hold
appears to be an issue with `concurrencyPolicy` default
commented in [Jira](https://issues.redhat.com/browse/OCPBUGS-9070?focusedId=22132293&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-22132293)

evakhoni commented:
follow up the bug ticket, since results are expected, lets consider this pre-merge verified
/unhold
/label qe-approved

Davoska commented:
/retest

petr-muller commented:
Looks safe to override to me

Davoska commented:
/override ci/prow/e2e-agnostic-ovn-upgrade-into-change

The `ci/prow/e2e-agnostic-ovn-upgrade-into-change` job has already passed two times for this code change.

The currently failing run's cause doesn't seem to be related to the PR.
